runtime/onert/core/src/util/shapeinf/Gather.cc

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2020 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "util/ShapeInference.h"
+
+namespace onert
+{
+namespace shape_inference
+{
+
+ir::Shape GatherShapes(const ir::Shape &input_shape, const ir::Shape &indices_shape, int axis,
+                       int rank)
+{
+  ir::Shape out_shape;
+  const int indices_rank = indices_shape.rank();
+  for (int idx = 0; idx < rank; ++idx)
+  {
+    if (idx == axis)
+    {
+      for (int indices_idx = 0; indices_idx < indices_rank; indices_idx++)
+      {
+        out_shape.append(indices_shape.dim(indices_idx));
+      }
+    }
+    else
+    {
+      out_shape.append(input_shape.dim(idx));
+    }
+  }
+
+  return out_shape;
+}
+
+void StaticInferer::visit(const ir::operation::Gather &op)
+{
+  const auto input_idx{op.getInputs().at(ir::operation::Gather::Input::INPUT)};
+  const auto &input = _operands.at(input_idx);
+
+  // get mutable output operand
+  const auto output_idx = op.getOutputs().at(0);
+  ir::Operand &output = _operands.at(output_idx);
+
+  // if input is dynamic, output also becomes dynamic
+  if (input.info().isDynamic())
+  {
+    output.info().setDynamic();
+    return;
+  }
+
+  const auto rank = input.info().shape().rank();
+  const auto axis = ((op.param().axis < 0) ? rank + op.param().axis : op.param().axis);
+
+  assert(0 <= axis && axis < rank);
+
+  const auto indices_idx{op.getInputs().at(ir::operation::Gather::Input::INDICES)};
+  const auto &indices = _operands.at(indices_idx);
+  // re-sizing output shape
```

Comment:
I wonder if `indices` can be an output of other op. If so, doesn't `indices` affect the decision of dynamic/static aspect of output?
If we want `indices` to be always const, we'd better add some `throw exception if indeces is not const` statement somewhere.

Reply:
`indices` itself could be an output of other op. However, If shape of `indices` is constant, the output dimension of `Gather` could be calculated.